File: src/cli/tui/widgets/modal.py
"""Modal widget — generic centered dialog with title and content."""

from __future__ import annotations

from textual.app import ComposeResult
from textual.containers import Center, Vertical
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Static

from cli.tui.theme import ACCENT_AMBER, BG_SECONDARY, TEXT_PRIMARY


class NiuuModal(Widget):
    """Generic centered modal dialog.

    Set ``visible`` reactive to show/hide.
    """

    DEFAULT_CSS = f"""
    NiuuModal {{
        align: center middle;
        display: none;
    }}
    NiuuModal.visible {{
        display: block;
        layer: modal;
    }}
    NiuuModal #modal-container {{
        width: 60;
        max-height: 20;
        background: {BG_SECONDARY};
        border: round {ACCENT_AMBER};
        padding: 1 2;
    }}
    NiuuModal #modal-title {{
        text-align: center;
        text-style: bold;
        color: {ACCENT_AMBER};
        height: 1;
        margin-bottom: 1;
    }}
    NiuuModal #modal-body {{
        height: auto;
        color: {TEXT_PRIMARY};
    }}
    """

    is_visible: reactive[bool] = reactive(False)

    def __init__(
        self,
        title: str = "",
        content: str = "",
        *,
        name: str | None = None,
        id: str | None = None,
        classes: str | None = None,
    ) -> None:
        super().__init__(name=name, id=id, classes=classes)
        self._title = title
        self._content = content

    @property
    def title(self) -> str:
        return self._title

    @property
    def content(self) -> str:
        return self._content

    def compose(self) -> ComposeResult:
        with Center():
            with Vertical(id="modal-container"):
                yield Static(self._title, id="modal-title")
                yield Static(self._content, id="modal-body")

    def watch_is_visible(self, value: bool) -> None:
        if value:
            self.add_class("visible")
        else:
            self.remove_class("visible")

    def show(self, title: str | None = None, content: str | None = None) -> None:
        if title is not None:
            self._title = title
            try:
                self.query_one("#modal-title", Static).update(title)
            except Exception:
                pass
        if content is not None:
            self._content = content
            try:
                self.query_one("#modal-body", Static).update(content)
            except Exception:
                pass
        self.is_visible = True

    def hide(self) -> None:
        self.is_visible = False

File: cli/internal/tui/components/modal.go
package components

import (
	"charm.land/lipgloss/v2"
	tui "github.com/niuulabs/volundr/cli/internal/tui"
)

// Modal renders a centered overlay dialog.
type Modal struct {
	Title   string
	Content string
	Width   int
	Height  int
	Visible bool
}

// NewModal creates a new modal with the given title.
func NewModal(title string) Modal {
	return Modal{
		Title:   title,
		Width:   60,
		Height:  20,
		Visible: false,
	}
}

// View renders the modal if visible.
func (m Modal) View(termWidth, termHeight int) string {
	if !m.Visible {
		return ""
	}

	theme := tui.DefaultTheme

	titleStyle := lipgloss.NewStyle().
		Foreground(theme.AccentAmber).
		Bold(true).
		MarginBottom(1)

	contentStyle := lipgloss.NewStyle().
		Foreground(theme.TextPrimary)

	modal := lipgloss.NewStyle().
		Background(theme.BgSecondary).
		Border(lipgloss.RoundedBorder()).
		BorderForeground(theme.AccentAmber).
		Padding(1, 2).
		Width(m.Width).
		MaxHeight(m.Height).
		Render(titleStyle.Render(m.Title) + "\n" + contentStyle.Render(m.Content))

	return lipgloss.Place(
		termWidth, termHeight,
		lipgloss.Center, lipgloss.Center,
		modal,
	)
}
